[css-grid] Use synthesizedBaseline to synthesize the correct baseline of grid items in the column axis.

sammygill opened the pull request:
#### 2688e1f33c192b78b77f2a49f787171cce7f6db1
<pre>
[css-grid] Use synthesizedBaseline to synthesize the correct baseline of grid items in the column axis.
<a href="https://bugs.webkit.org/show_bug.cgi?id=263600">https://bugs.webkit.org/show_bug.cgi?id=263600</a>
rdar://problem/117424263

Reviewed by Alan Baradlay.

Currently GridBaselineAlignment::ascentForChild does not correctly
syntheisze the central baseline in certain situations. It appears that
the current code attempts to only synthesize the alphabetic baseline.
Accoring to the spec, alignment-baseline: baseline, which is the initial
value, refers to the box&apos;s parent&apos;s dominant-baseline, which is the
initial value of &quot;auto,&quot; since we do not implement it yet either.
The description for &quot;dominant-baseline: auto,&quot; says:

Equivalent to alphabetic in horizontal writing modes and in vertical
writing modes when text-orientation is sideways. Equivalent to central
in vertical writing modes when text-orientation is mixed or upright.

Instead of attempting to synthesize the baseline within grid layout
code, we can use synthesizedBaseline in RenderBox which synthesizes
a baseline value depending on the arguments it is given. This function
will compute the central baseline of the box in the scenarios that are
described above in the property value&apos;s description. We perform this
logic only when we are aligning in the column axis and fallback to
the old behavior for the row axis. Attempting to change this logic for
both at this time causes regressions so a closer look will be needed in
order to address that case in a separate patch.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-008-expected.txt:
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::GridBaselineAlignment::ascentForChild const):

Canonical link: <a href="https://commits.webkit.org/269857@main">https://commits.webkit.org/269857@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bbd17d940133276dc4c748685d63f9bae538e202

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1784 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24790 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21850 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3448 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24205 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22404 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23913 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1406 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20479 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26424 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1192 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21386 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27655 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21657 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21654 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25421 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1092 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18795 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1074 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5696 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1506 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1387 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/GridBaselineAlignment.cpp:
```diff
@@ -80,15 +80,33 @@ LayoutUnit GridBaselineAlignment::logicalAscentForChild(const RenderBox& child,
+
+    if (baselineAxis == GridAxis::GridColumnAxis) {
+        ASSERT(child.parentStyle());
+        const auto* parentStyle = child.parentStyle();
```
(nit; const is redundant)

sammygill replied:
Removed

alanbaradlay commented on Source/WebCore/rendering/GridBaselineAlignment.cpp:
```diff
@@ -80,15 +80,33 @@ LayoutUnit GridBaselineAlignment::logicalAscentForChild(const RenderBox& child,
+
+        if (!isParallelToBaselineAxisForChild(child, baselineAxis))
+            return (parentStyle ? synthesizedBaseline(child, *child.parentStyle(), alignmentContextDirection(), BaselineSynthesisEdge::BorderBox) : 0_lu) + margin;
+        std::optional<LayoutUnit> ascent = position == ItemPosition::Baseline ? child.firstLineBaseline() : child.lastLineBaseline();
```
auto?

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/GridBaselineAlignment.cpp:
```diff
@@ -80,15 +80,33 @@ LayoutUnit GridBaselineAlignment::logicalAscentForChild(const RenderBox& child,
+            return (parentStyle ? synthesizedBaseline(child, *child.parentStyle(), alignmentContextDirection(), BaselineSynthesisEdge::BorderBox) : 0_lu) + margin;
+        baseline = ascent.value();
+    } else {
+        std::optional<LayoutUnit> computedBaselineValue = position == ItemPosition::Baseline ? child.firstLineBaseline() : child.lastLineBaseline();
```
auto

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/GridBaselineAlignment.cpp:
```diff
@@ -80,15 +80,33 @@ LayoutUnit GridBaselineAlignment::logicalAscentForChild(const RenderBox& child,
+        baseline = ascent.value();
+    } else {
+        std::optional<LayoutUnit> computedBaselineValue = position == ItemPosition::Baseline ? child.firstLineBaseline() : child.lastLineBaseline();
+        baseline = isParallelToBaselineAxisForChild(child, baselineAxis) ? computedBaselineValue.value_or(LayoutUnit(-1)) : LayoutUnit(-1);
```
we should refrain from using magic values if possible

sammygill replied:
Added a const variable to the top of the method

sammygill committed:
[css-grid] Use synthesizedBaseline to synthesize the correct baseline of grid items in the column axis.
https://bugs.webkit.org/show_bug.cgi?id=263600
rdar://problem/117424263

Reviewed by Alan Baradlay.

Currently GridBaselineAlignment::ascentForChild does not correctly
syntheisze the central baseline in certain situations. It appears that
the current code attempts to only synthesize the alphabetic baseline.
Accoring to the spec, alignment-baseline: baseline, which is the initial
value, refers to the box's parent's dominant-baseline, which is the
initial value of "auto," since we do not implement it yet either.
The description for "dominant-baseline: auto," says:

Equivalent to alphabetic in horizontal writing modes and in vertical
writing modes when text-orientation is sideways. Equivalent to central
in vertical writing modes when text-orientation is mixed or upright.

Instead of attempting to synthesize the baseline within grid layout
code, we can use synthesizedBaseline in RenderBox which synthesizes
a baseline value depending on the arguments it is given. This function
will compute the central baseline of the box in the scenarios that are
described above in the property value's description. We perform this
logic only when we are aligning in the column axis and fallback to
the old behavior for the row axis. Attempting to change this logic for
both at this time causes regressions so a closer look will be needed in
order to address that case in a separate patch.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-008-expected.txt:
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::GridBaselineAlignment::ascentForChild const):

Canonical link: https://commits.webkit.org/269857@main
